lib/ansible/module_utils/identity/keycloak/keycloak.py

```diff
@@ -68,49 +68,60 @@ def camel(words):
+    def __init__(self, base_url, validate_certs, auth_realm, client_id,
+                 auth_username, auth_password, client_secret):
+        self.validate_certs = validate_certs
+        self.auth_url = URL_TOKEN.format(url=base_url, realm=auth_realm)
+        temp_payload = {
+            'grant_type': 'password',
+            'client_id': client_id,
+            'client_secret': client_secret,
+            'username': auth_username,
+            'password': auth_password,
+        }
         # Remove empty items, for instance missing client_secret
-        payload = dict((k, v) for k, v in payload.items() if v is not None)
+        self.payload = dict(
+            (k, v) for k, v in temp_payload.items() if v is not None)
+        self.header = {}
+        self.refresh_token()
 
+    def refresh_token(self):
         try:
-            r = json.load(open_url(auth_url, method='POST',
-                                   validate_certs=self.validate_certs, data=urlencode(payload)))
+            r = json.load(open_url(self.auth_url, method='POST',
+                                   validate_certs=self.validate_certs,
+                                   data=urlencode(self.payload)))
         except ValueError as e:
-            self.module.fail_json(msg='API returned invalid JSON when trying to obtain access token from %s: %s'
-                                      % (auth_url, str(e)))
+            raise KeycloakError(
+                'API returned invalid JSON when trying to obtain access token from %s: %s'
+                % (self.auth_url, str(e)))
         except Exception as e:
-            self.module.fail_json(msg='Could not obtain access token from %s: %s'
-                                      % (auth_url, str(e)))
+            raise KeycloakError('Could not obtain access token from %s: %s'
+                                % (self.auth_url, str(e)))
 
-        if 'access_token' in r:
-            self.token = r['access_token']
-            self.restheaders = {'Authorization': 'Bearer ' + self.token,
-                                'Content-Type': 'application/json'}
+        try:
+            self.header = {
+                'Authorization': 'Bearer ' + r['access_token'],
+                'Content-Type': 'application/json'
+            }
+        except KeyError:
+            raise KeycloakError(
+                'Could not obtain access token from %s' % self.auth_url)
 
-        else:
-            self.module.fail_json(msg='Could not obtain access token from %s' % auth_url)
+
+class KeycloakAPI(object):
+    """ Keycloak API access; Keycloak uses OAuth 2.0 to protect its API, an access token for which
+        is obtained through OpenID connect
+    """
+    def __init__(self, module, connection_header):
+        self.module = module
+        self.baseurl = self.module.params.get('auth_keycloak_url')
+        self.validate_certs = self.module.params.get('validate_certs')
+        self.restheaders = connection_header
```

Comment:
Maybe this could use `self.restheaders = connection_header.header` to avoid changing all the calls bellow to `headers=self.restheaders.header`. Especially since I don't think the object's other fields are used again.

Reply:
I give the whole object in order to use the `refresh_token` later.

But as I think about it, it seems useless, because the token is generated at the beginning of every module.

No I don't use the other fields after. I think a function is enough.